### DISCUSSION TITLE
CloudWatch: Use union of ListMetrics API result and hard codes metrics

### DISCUSSION THREAD

**Discussion by sunker** [grafana/grafana#41814] [Data Sources]
**What would you like to be added**:
The metrics, namespaces and dimension keys that are shown in the query editor in Grafana should be a union of the result from the [ListMetrics API](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_ListMetrics.html) and the [hard coded list](https://github.com/grafana/grafana/blob/master/pkg/tsdb/cloudwatch/metric_find_query.go#L44-L234) of namespaces, metrics and dimension keys that we already have in Grafana. 

**Why is this needed**:
At the moment, namespaces, metrics and dimension keys are  [hard coded](https://github.com/grafana/grafana/blob/master/pkg/tsdb/cloudwatch/metric_find_query.go#L44-L234) in Grafana. The problem with that is that each time a new namespace, metric or dimension is added in AWS, this hard coded list needs to be updated manually. Also, this doesn't take custom namespaces into consideration, so at the moment they have to be added manually in the [data source configuration page](https://grafana.com/docs/grafana/latest/datasources/cloudwatch/#adding-the-data-source). 

The aforementioned problem would be partly solved by using [ListMetrics API](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_ListMetrics.html). However, the ListMetrics API doesn't return information about metrics if those metrics haven't reported data in the past two weeks. So if we switched over to only use the API, we would potentially change the query editor behavior for our users. Therefore, we should present a union of the hard coded metrics and the result from ListMetrics API in Grafana.

The ListMetrics API is part of the [paid tier](https://aws.amazon.com/cloudwatch/pricing/#Paid_tier), so we should not issue a new request to the API each time the namespace, metrics or dimension key field is clicked on in the query editor or is requested using a template variable query. Instead, the result should be cached and used for some time. I suggest 1 hour. 

This is related to #28382, #23134, #22215, #26505

Feedback appreciated @mtanda @chrs-myrs @mchene

**Comment by mtanda** [comment]
@sunker We should consider about large environment, which have many metrics in one AWS account, or have many AWS accounts.

I think there are two approaches to implement this feature.

- (on-demand) call ListMetrics API when opening query editor
- (scheduled) call ListMetrics API periodically in Grafana backend process, query editor refer the cached response

Both approaches have some difficulties.

**Problem1: query editor UX could be bad due to slow response**

To get all MetricName in the Namespace by ListMetrics API, Grafana need to get all metrics in the Namespace.
The response time could be too slow for on-demand use case.

I check metrics count in one of our large AWS account.
```
$ aws cloudwatch list-metrics --namespace AWS/EC2 | jq -c '.Metrics[]' | wc -l
54659
```
One ListMetrics API call respond 500 metrics, so the command call API 100 times or more.

```
$ aws cloudwatch list-metrics --namespace AWS/EC2 | jq -c '.Metrics[].MetricName' | sort -u | wc -l
19
```
The unique MetricName count is only 19...

If EC2 instances are launched by Auto Scaling, the metrics count tends to increase.
The ListMetrics `RecentlyActive` parameter may reduce the count, but it may drop MetricName which is not recorded recently.

**Problem 2: Grafana need to call ListMetrics API for all regions / accounts which are accessible**

Grafana may have multiple CloudWatch datasource, and one datasource can access multiple regions / multiple accounts.
If Grafana need to collect all MetricName in those regions / accounts, Grafana need to call ListMetrics API so many times.

We need policy to select target region / account.

**Comment by chrs-myrs** [comment]
Hi @sunker, thanks for picking this up, here are my thoughts.

I think your concept of using a union of the lists makes perfect sense.  If it would be better to do something minimal there could just be a button on the data source page that refreshed the existing list from the ListMetrics call.  That would be a huge step forward for us as currently updating the Namespaces has to be done in our provisioning tool.  Not sure if the list would then need to be detached from the provisioned datasource config (or a second list managed in a similar way to your suggestion) so that it is not reset when provisioning.

In terms of automating this update, having a scheduled update would be good, but I feel this would need to be an optional setting exposed to users, since random updating in the background makes it hard to debug data problems.  Exposing an API endpoint would be enough, as we could schedule this externally if we wanted, but personally I wouldn't bother, we do create new namespaces regularly, and have to update these across 11 accounts, but clicking an button would be fine for this, especially if there were an info popup against the namespace drop-down in the query editor that linked directly to the relevant datasource config page.

**Comment by sunker** [comment]
Thanks for feedback @mtanda and @chrs-myrs! 

I would really like to avoid having scheduled api calls in our backend. I think these should be loaded on demand and then cached. Then it becomes obvious to the user what it is that is taking time. 

> I check metrics count in one of our large AWS account.
> 
> ```
> $ aws cloudwatch list-metrics --namespace AWS/EC2 | jq -c '.Metrics[]' | wc -l
> 54659
> ```

I had no clue it could be that amount of metrics. Since that's the case, I don't think it's feasible to use the list metrics api at all. What we would need is: 
* a new api that could list namespaces (and nothing else)
* a new api that could list metric names per namespace (and nothing else)
* a new api that could list dimension keys per namespace and metric (and nothing else)

**Comment by chrs-myrs** [comment]
Ah that's a lot of metrics, we have a lot of metrics and cloudwatch is one of our bigger costs, but we still only have 2,500 in total (75 in the EC2 namespace).

Will feed back to our AWS account manager that this is a bit of an oversight not having a CLI command for list namespaces, if anyone else has a way of raising a feature request feel free

**Comment by sunker** [comment]
I appreciate that @chrs-myrs! I'll also report back to our contact person at AWS.

**Comment by peycho** [comment]
We have even more. Don't ask why.

```
~$ aws cloudwatch list-metrics --namespace AWS/EC2 | jq -c '.Metrics[]' | wc -l
421012
```

**Comment by sarahzinger** [comment]
As part of our work moving from [discussions to issues](https://github.com/grafana/grafana/discussions/73424), I am closing this discussion in favor of [this github issue](https://github.com/grafana/grafana/issues/77963). Feel free to add any other suggestions or ideas for the best way to handle our dependency on hard coded metrics there. Thanks!!
